src/bm_sim/P4Objects.cpp

```diff
@@ -1002,19 +1002,127 @@ P4Objects::init_actions(const Json::Value &cfg_root) {
         {"valid", MatchKeyParam::Type::VALID},
         {"range", MatchKeyParam::Type::RANGE} };
 
+  auto it = map_name_to_match_type.find(name);
+  if (it == map_name_to_match_type.end())
+    throw json_exception(EFormat() << "Invalid match type: '" << name << "'");
+  return it->second;
+}
+
+std::vector<MatchKeyParam> parse_match_key(
+    const Json::Value &cfg_match_key, const Json::Value &cfg_ref_match_key) {
+  std::vector<MatchKeyParam> match_key;
+
+  if (cfg_match_key.size() != cfg_ref_match_key.size()) {
+    throw json_exception(
+        EFormat() << "Invalid number of fields in match key, expected "
+                  << cfg_ref_match_key.size() << " but got "
+                  << cfg_match_key.size(),
+        cfg_match_key);
+  }
+
+  auto convert_valid_key = [](bool key) {
+    return key ? std::string("\x01", 1) : std::string("\x00", 1);
+  };
+
+  auto transform_hexstr = [](const std::string hexstr) {
+    ByteContainer bc(hexstr);
+    return std::string(bc.data(), bc.size());
+  };
+
+  for (size_t i = 0; i < cfg_match_key.size(); i++) {
+    // cannot use [i] directly because of ambiguous operator overloading in
+    // jsoncpp code
+    auto index = static_cast<Json::ArrayIndex>(i);
+    const auto &cfg_f = cfg_match_key[index];
+    const auto match_type_str = cfg_f["match_type"].asString();
+    const auto ref_match_type_str =
+        cfg_ref_match_key[index]["match_type"].asString();
+
+    if (match_type_str != ref_match_type_str) {
+      throw json_exception(
+          EFormat() << "Invalid match type for field #" << i
+                    << " in match key, expected " << ref_match_type_str
+                    << " but got " << match_type_str,
+          cfg_match_key);
+    }
+
+    const auto match_type = match_name_to_match_type(match_type_str);
+    switch (match_type) {
+      case MatchKeyParam::Type::EXACT:
+        match_key.emplace_back(match_type,
+                               transform_hexstr(cfg_f["key"].asString()));
+        break;
+      case MatchKeyParam::Type::LPM:
+        match_key.emplace_back(match_type,
+                               transform_hexstr(cfg_f["key"].asString()),
+                               cfg_f["prefix_length"].asInt());
+        break;
+      case MatchKeyParam::Type::TERNARY:
+        match_key.emplace_back(match_type,
+                               transform_hexstr(cfg_f["key"].asString()),
+                               transform_hexstr(cfg_f["mask"].asString()));
+        break;
+      case MatchKeyParam::Type::VALID:
+        match_key.emplace_back(match_type,
+                               convert_valid_key(cfg_f["key"].asBool()));
+        break;
+      case MatchKeyParam::Type::RANGE:
+        match_key.emplace_back(match_type,
+                               transform_hexstr(cfg_f["start"].asString()),
+                               transform_hexstr(cfg_f["end"].asString()));
+        break;
+    }
+  }
+  return match_key;
+}
+
```

Comment:
not sure I understand this, I do check that the format of the match key is correct